{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ECCITE_seq Dataset Download and Preprocessing "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Preprocessing data code from repository suinleelab **contrastiveVI**: https://github.com/suinleelab/contrastiveVI/blob/main/contrastive_vi/data/datasets/papalexi_2021.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The link to download the data is in the code. The ECCITE_seq data is from Gene Expression Omnibus (GEO) public genomics data repository (https://www.ncbi.nlm.nih.gov/geo/info/download.html) The license of the organization can be found here: https://catalog.data.gov/dataset/gene-expression-omnibus-geo. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The python **anndata** package is found here: https://github.com/scverse/anndata?tab=BSD-3-Clause-1-ov-file\n",
    "\n",
    "The **scanpy** package is found here: https://github.com/scverse/scanpy\n",
    "\n",
    "The **requests** package is found here: https://github.com/psf/requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Download, read, and preprocess Papalexi et al. (2021) expression data.\n",
    "\n",
    "Single-cell expression data from Papalexi et al. Characterizing the molecular regulation\n",
    "of inhibitory immune checkpoints with multimodal single-cell screens. (Nature Genetics\n",
    "2021)\n",
    "\"\"\"\n",
    "import os\n",
    "import shutil\n",
    "\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from anndata import AnnData\n",
    "\n",
    "import requests\n",
    "\n",
    "\n",
    "def download_binary_file(file_url: str, output_path: str) -> None:\n",
    "    \"\"\"\n",
    "    Download binary data file from a URL.\n",
    "\n",
    "    Args:\n",
    "    ----\n",
    "        file_url: URL where the file is hosted.\n",
    "        output_path: Output path for the downloaded file.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "        None.\n",
    "    \"\"\"\n",
    "    request = requests.get(file_url)\n",
    "    with open(output_path, \"wb\") as f:\n",
    "        f.write(request.content)\n",
    "    print(f\"Downloaded data from {file_url} at {output_path}\")\n",
    "\n",
    "\n",
    "\n",
    "def download_papalexi_2021(output_path: str) -> None:\n",
    "    \"\"\"\n",
    "    Download Papalexi et al. 2021 data from the hosting URLs.\n",
    "\n",
    "    Args:\n",
    "    ----\n",
    "        output_path: Output path to store the downloaded and unzipped\n",
    "        directories.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "        None. File directories are downloaded to output_path.\n",
    "    \"\"\"\n",
    "\n",
    "    counts_data_url = (\n",
    "        \"https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE153056&format=file\"\n",
    "    )\n",
    "    data_output_filename = os.path.join(output_path, \"GSE153056_RAW.tar\")\n",
    "    download_binary_file(counts_data_url, data_output_filename)\n",
    "    shutil.unpack_archive(data_output_filename, output_path)\n",
    "\n",
    "    metadata_url = (\n",
    "        \"https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE153056&\"\n",
    "        \"format=file&file=GSE153056_ECCITE_metadata.tsv.gz\"\n",
    "    )\n",
    "    metadata_filename = os.path.join(output_path, metadata_url.split(\"=\")[-1])\n",
    "    download_binary_file(metadata_url, metadata_filename)\n",
    "\n",
    "\n",
    "def read_papalexi_2021(file_directory: str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Read the expression data for Papalexi et al. 2021 in the given directory.\n",
    "\n",
    "    Args:\n",
    "    ----\n",
    "        file_directory: Directory containing Papalexi et al. 2021 data.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "        A pandas dataframe, with each column representing a cell\n",
    "        and each row representing a gene feature.\n",
    "    \"\"\"\n",
    "\n",
    "    matrix = pd.read_csv(\n",
    "        os.path.join(file_directory, \"GSM4633614_ECCITE_cDNA_counts.tsv.gz\"),\n",
    "        sep=\"\\t\",\n",
    "        index_col=0,\n",
    "    )\n",
    "    return matrix\n",
    "\n",
    "\n",
    "def preprocess_papalexi_2021(download_path: str, n_top_genes: int) -> AnnData:\n",
    "    \"\"\"\n",
    "    Preprocess expression data from Papalexi et al. 2021.\n",
    "\n",
    "    Args:\n",
    "    ----\n",
    "        download_path: Path containing the downloaded Papalexi et al. 2021 data files.\n",
    "        n_top_genes: Number of most variable genes to retain.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "        An AnnData object containing single-cell expression data. The layer\n",
    "        \"count\" contains the count data for the most variable genes. The .X\n",
    "        variable contains the normalized and log-transformed data for the most variable\n",
    "        genes. A copy of data with all genes is stored in .raw.\n",
    "    \"\"\"\n",
    "\n",
    "    df = read_papalexi_2021(download_path)\n",
    "\n",
    "    # Switch dataframe from gene rows and cell columns to cell rows and gene columns\n",
    "    df = df.transpose()\n",
    "\n",
    "    metadata = pd.read_csv(\n",
    "        os.path.join(download_path, \"GSE153056_ECCITE_metadata.tsv.gz\"),\n",
    "        sep=\"\\t\",\n",
    "        index_col=0,\n",
    "    )\n",
    "\n",
    "    metadata.to_csv(\"eccite_metadata.csv\")\n",
    "\n",
    "    # Note: By initializing the anndata object from a dataframe, variable names\n",
    "    # are automatically stored in adata.var\n",
    "    adata = AnnData(df)\n",
    "    adata.obs = metadata\n",
    "\n",
    "    # Protein measurements also collected as part of CITE-Seq\n",
    "    protein_counts_df = pd.read_csv(\n",
    "        os.path.join(download_path, \"GSM4633615_ECCITE_ADT_counts.tsv.gz\"),\n",
    "        sep=\"\\t\",\n",
    "        index_col=0,\n",
    "    )\n",
    "\n",
    "    # Switch dataframe from protein rows and cell columns to cell rows and protein\n",
    "    # columns\n",
    "    protein_counts_df = protein_counts_df.transpose()\n",
    "\n",
    "    # Storing protein counts in an obsm field as expected by totalVI\n",
    "    # (see https://docs.scvi-tools.org/en/stable/tutorials/notebooks/totalVI.html\n",
    "    # for an example). Since `protein_counts_df` is annotated with protein names,\n",
    "    # our obsm field will retain them as well.\n",
    "    # adata.obsm[constants.PROTEIN_EXPRESSION_KEY] = protein_counts_df\n",
    "\n",
    "    adata.layers[\"count\"] = adata.X.copy()\n",
    "    sc.pp.normalize_total(adata)\n",
    "    sc.pp.log1p(adata)\n",
    "    adata.raw = adata\n",
    "    sc.pp.highly_variable_genes(\n",
    "        adata, flavor=\"seurat_v3\", n_top_genes=n_top_genes, layer=\"count\", subset=True\n",
    "    )\n",
    "    adata = adata[adata.layers[\"count\"].sum(1) != 0]  # Remove cells with all zeros.\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "root_data_path = \"...\"   # local computer data path to downloaad the dataset. \n",
    "\n",
    "download_papalexi_2021(root_data_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_top_genes = 1000 \n",
    "\n",
    "data = preprocess_papalexi_2021(root_data_path, n_top_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process to Get Foreground and Background  dataset\n",
    "Foreground is treatment, and background is control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2386\n",
      "                    orig.ident  nCount_RNA  nFeature_RNA  nCount_HTO  \\\n",
      "l1_AAACGGGAGACAGAGA      Lane1       14941          3918          97   \n",
      "l1_AAAGCAACAGTGGAGT      Lane1        6592          2564          41   \n",
      "l1_AAAGTAGAGGATCGCA      Lane1        9783          3334          90   \n",
      "l1_AACGTTGTCATGCAAC      Lane1       12147          3233          31   \n",
      "l1_AAGGAGCCATTAGGCT      Lane1       10656          3259          91   \n",
      "...                        ...         ...           ...         ...   \n",
      "l8_TTTCCTCTCAGCAACT      Lane8       16790          4298         116   \n",
      "l8_TTTGGTTGTTCCTCCA      Lane8        5093          1756         133   \n",
      "l8_TTTGGTTTCGACCAGC      Lane8        5837          2302          87   \n",
      "l8_TTTGTCAAGTTACGGG      Lane8       22441          4624         170   \n",
      "l8_TTTGTCAGTGACGGTA      Lane8       17553          3787         528   \n",
      "\n",
      "                     nFeature_HTO  nCount_GDO  nFeature_GDO  nCount_ADT  \\\n",
      "l1_AAACGGGAGACAGAGA             5         124           111         487   \n",
      "l1_AAAGCAACAGTGGAGT             5        1278           111         458   \n",
      "l1_AAAGTAGAGGATCGCA             4         162           111         561   \n",
      "l1_AACGTTGTCATGCAAC             3         365           111         458   \n",
      "l1_AAGGAGCCATTAGGCT             2        1101           111         470   \n",
      "...                           ...         ...           ...         ...   \n",
      "l8_TTTCCTCTCAGCAACT             3         220           111         394   \n",
      "l8_TTTGGTTGTTCCTCCA             5         164           111         226   \n",
      "l8_TTTGGTTTCGACCAGC             2         164           111         471   \n",
      "l8_TTTGTCAAGTTACGGG             2         147           111         212   \n",
      "l8_TTTGTCAGTGACGGTA             3         870           111        3042   \n",
      "\n",
      "                     nFeature_ADT  percent.mito  ... guide_ID guide_ID.global  \\\n",
      "l1_AAACGGGAGACAGAGA             4      3.379961  ...     NTg1         Singlet   \n",
      "l1_AAAGCAACAGTGGAGT             4      5.552184  ...     NTg4         Singlet   \n",
      "l1_AAAGTAGAGGATCGCA             4      2.545232  ...     NTg1         Singlet   \n",
      "l1_AACGTTGTCATGCAAC             4      3.531736  ...     NTg5         Singlet   \n",
      "l1_AAGGAGCCATTAGGCT             4      2.449324  ...     NTg8         Singlet   \n",
      "...                           ...           ...  ...      ...             ...   \n",
      "l8_TTTCCTCTCAGCAACT             4      3.323407  ...     NTg2         Singlet   \n",
      "l8_TTTGGTTGTTCCTCCA             4      5.262124  ...     NTg5         Singlet   \n",
      "l8_TTTGGTTTCGACCAGC             4      5.105362  ...     NTg9         Singlet   \n",
      "l8_TTTGTCAAGTTACGGG             4      3.435676  ...    NTg10         Singlet   \n",
      "l8_TTTGTCAGTGACGGTA             4      2.159175  ...     NTg4         Singlet   \n",
      "\n",
      "                    gene con  NT crispr replicate   S.Score G2M.Score Phase  \n",
      "l1_AAACGGGAGACAGAGA   NT  tx  NT     NT      rep1 -0.209984 -0.584292    G1  \n",
      "l1_AAAGCAACAGTGGAGT   NT  tx  NT     NT      rep1  0.043371 -0.232169     S  \n",
      "l1_AAAGTAGAGGATCGCA   NT  tx  NT     NT      rep1 -0.028834 -0.289386    G1  \n",
      "l1_AACGTTGTCATGCAAC   NT  tx  NT     NT      rep1 -0.145527 -0.444884    G1  \n",
      "l1_AAGGAGCCATTAGGCT   NT  tx  NT     NT      rep1  0.034585 -0.378017     S  \n",
      "...                  ...  ..  ..    ...       ...       ...       ...   ...  \n",
      "l8_TTTCCTCTCAGCAACT   NT  tx  NT     NT      rep3 -0.138019 -0.696598    G1  \n",
      "l8_TTTGGTTGTTCCTCCA   NT  tx  NT     NT      rep2 -0.021565 -0.122678    G1  \n",
      "l8_TTTGGTTTCGACCAGC   NT  tx  NT     NT      rep2  0.140815  0.040613     S  \n",
      "l8_TTTGTCAAGTTACGGG   NT  tx  NT     NT      rep2 -0.203834 -0.910296    G1  \n",
      "l8_TTTGTCAGTGACGGTA   NT  tx  NT     NT      rep2 -0.153514 -0.633978    G1  \n",
      "\n",
      "[2386 rows x 24 columns]\n"
     ]
    }
   ],
   "source": [
    "control = data[data.obs[\"crispr\"] == \"NT\"]\n",
    "m, k = control.shape\n",
    "print(m)\n",
    "print(control.obs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "NT stands for Non-targeting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18343\n",
      "                    orig.ident  nCount_RNA  nFeature_RNA  nCount_HTO  \\\n",
      "l1_AAACCTGAGCCAGAAC      Lane1       17207          3942          99   \n",
      "l1_AAACCTGAGTGGACGT      Lane1        9506          2948          35   \n",
      "l1_AAACCTGCATGAGCGA      Lane1       15256          4258          66   \n",
      "l1_AAACCTGTCTTGTCAT      Lane1        5135          1780          22   \n",
      "l1_AAACGGGAGAACAACT      Lane1        9673          2671          99   \n",
      "...                        ...         ...           ...         ...   \n",
      "l8_TTTGTCAGTAGGCATG      Lane8        9446          3240         111   \n",
      "l8_TTTGTCAGTCACTTCC      Lane8       20863          4884         294   \n",
      "l8_TTTGTCAGTTCCACAA      Lane8       15106          4185         154   \n",
      "l8_TTTGTCATCACGCATA      Lane8       11209          3204         132   \n",
      "l8_TTTGTCATCTGTACGA      Lane8       23088          5073          95   \n",
      "\n",
      "                     nFeature_HTO  nCount_GDO  nFeature_GDO  nCount_ADT  \\\n",
      "l1_AAACCTGAGCCAGAAC             4         576           111         801   \n",
      "l1_AAACCTGAGTGGACGT             5         190           111         545   \n",
      "l1_AAACCTGCATGAGCGA             4         212           111         344   \n",
      "l1_AAACCTGTCTTGTCAT             3         243           111         539   \n",
      "l1_AAACGGGAGAACAACT             5         198           111        1053   \n",
      "...                           ...         ...           ...         ...   \n",
      "l8_TTTGTCAGTAGGCATG             3         118           111         178   \n",
      "l8_TTTGTCAGTCACTTCC             4         290           111         398   \n",
      "l8_TTTGTCAGTTCCACAA             6         267           111         212   \n",
      "l8_TTTGTCATCACGCATA             3         202           111         258   \n",
      "l8_TTTGTCATCTGTACGA             4         234           111         584   \n",
      "\n",
      "                     nFeature_ADT  percent.mito  ... guide_ID guide_ID.global  \\\n",
      "l1_AAACCTGAGCCAGAAC             4      2.295577  ...  STAT2g2         Singlet   \n",
      "l1_AAACCTGAGTGGACGT             4      4.512939  ...   CAV1g4         Singlet   \n",
      "l1_AAACCTGCATGAGCGA             4      4.116413  ...  STAT1g2         Singlet   \n",
      "l1_AAACCTGTCTTGTCAT             4      5.491723  ...   CD86g1         Singlet   \n",
      "l1_AAACGGGAGAACAACT             4      3.359868  ...   IRF7g2         Singlet   \n",
      "...                           ...           ...  ...      ...             ...   \n",
      "l8_TTTGTCAGTAGGCATG             4      5.219140  ...   ETV7g1         Singlet   \n",
      "l8_TTTGTCAGTCACTTCC             4      5.469012  ...  CMTM6g1         Singlet   \n",
      "l8_TTTGTCAGTTCCACAA             4      2.661194  ...   ATF2g1         Singlet   \n",
      "l8_TTTGTCATCACGCATA             4      7.369078  ...   CAV1g2         Singlet   \n",
      "l8_TTTGTCATCTGTACGA             4      2.191615  ...  STAT3g1         Singlet   \n",
      "\n",
      "                      gene con       NT     crispr replicate   S.Score  \\\n",
      "l1_AAACCTGAGCCAGAAC  STAT2  tx  STAT2g2  Perturbed      rep1 -0.252716   \n",
      "l1_AAACCTGAGTGGACGT   CAV1  tx   CAV1g4  Perturbed      rep1 -0.123802   \n",
      "l1_AAACCTGCATGAGCGA  STAT1  tx  STAT1g2  Perturbed      rep1 -0.154633   \n",
      "l1_AAACCTGTCTTGTCAT   CD86  tx   CD86g1  Perturbed      rep1 -0.061262   \n",
      "l1_AAACGGGAGAACAACT   IRF7  tx   IRF7g2  Perturbed      rep1 -0.132188   \n",
      "...                    ...  ..      ...        ...       ...       ...   \n",
      "l8_TTTGTCAGTAGGCATG   ETV7  tx   ETV7g1  Perturbed      rep2 -0.091613   \n",
      "l8_TTTGTCAGTCACTTCC  CMTM6  tx  CMTM6g1  Perturbed      rep3 -0.323562   \n",
      "l8_TTTGTCAGTTCCACAA   ATF2  tx   ATF2g1  Perturbed      rep3 -0.191933   \n",
      "l8_TTTGTCATCACGCATA   CAV1  tx   CAV1g2  Perturbed      rep3 -0.134585   \n",
      "l8_TTTGTCATCTGTACGA  STAT3  tx  STAT3g1  Perturbed      rep3 -0.461981   \n",
      "\n",
      "                    G2M.Score Phase  \n",
      "l1_AAACCTGAGCCAGAAC -0.771844    G1  \n",
      "l1_AAACCTGAGTGGACGT -0.335672    G1  \n",
      "l1_AAACCTGCATGAGCGA -0.693982    G1  \n",
      "l1_AAACCTGTCTTGTCAT -0.035986    G1  \n",
      "l1_AAACGGGAGAACAACT -0.356173    G1  \n",
      "...                       ...   ...  \n",
      "l8_TTTGTCAGTAGGCATG -0.309941    G1  \n",
      "l8_TTTGTCAGTCACTTCC -0.797846    G1  \n",
      "l8_TTTGTCAGTTCCACAA -0.578331    G1  \n",
      "l8_TTTGTCATCACGCATA -0.503400    G1  \n",
      "l8_TTTGTCATCTGTACGA -1.077826    G1  \n",
      "\n",
      "[18343 rows x 24 columns]\n"
     ]
    }
   ],
   "source": [
    "treatment = data[data.obs[\"crispr\"] != \"NT\"]\n",
    "t, k = treatment.shape\n",
    "print(t)\n",
    "print(treatment.obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "background = control.to_df()\n",
    "foreground = treatment.to_df()\n",
    "\n",
    "background.to_csv(\"ECCITE_seq_back.csv\")\n",
    "foreground.to_csv(\"ECCITE_seq_fore.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
